packages/bundlers/default/src/DefaultBundler.js

```diff
@@ -1,191 +1,237 @@
 // @flow strict-local
 
-import type {BundleGroup, MutableBundle} from '@parcel/types';
+import type {Asset, Bundle} from '@parcel/types';
 
 import invariant from 'assert';
 import {Bundler} from '@parcel/plugin';
+import {md5FromString} from '@parcel/utils';
+import nullthrows from 'nullthrows';
 
 const OPTIONS = {
   minBundles: 1,
   minBundleSize: 30000,
   maxParallelRequests: 5
 };
 
-type BundleContext = {|
-  bundleGroup: BundleGroup,
-  bundle?: MutableBundle
-|};
-
 export default new Bundler({
-  bundle({assetGraph, bundleGraph}) {
-    // RULES:
-    // 1. If dep.isAsync or dep.isEntry, start a new bundle group.
-    // 2. If an asset is a different type than the current bundle, make a parallel bundle in the same bundle group.
-    // 3. If an asset is already in a parent bundle in the same entry point, exclude from child bundles.
-    // 4. If an asset is only in separate isolated entry points (e.g. workers, different HTML pages), duplicate it.
-    // 5. If the sub-graph from an asset is >= 30kb, and the number of parallel requests in the bundle group is < 5, create a new bundle containing the sub-graph.
-    // 6. If two assets are always seen together, put them in the same extracted bundle.
-
+  // RULES:
+  // 1. If dep.isAsync or dep.isEntry, start a new bundle group.
+  // 2. If an asset is a different type than the current bundle, make a parallel bundle in the same bundle group.
+  // 3. If an asset is already in a parent bundle in the same entry point, exclude from child bundles.
+  // 4. If an asset is only in separate isolated entry points (e.g. workers, different HTML pages), duplicate it.
+  // 5. If the sub-graph from an asset is >= 30kb, and the number of parallel requests in the bundle group is < 5, create a new bundle containing the sub-graph.
+  // 6. If two assets are always seen together, put them in the same extracted bundle
+
+  bundle({bundleGraph}) {
     // Step 1: create bundles for each of the explicit code split points.
-    assetGraph.traverse(
-      (node, context: ?BundleContext): ?BundleContext => {
-        if (node.type === 'dependency') {
-          let dep = node.value;
-
-          // Start a new bundle if this is an async dependency, or entry point.
-          if (dep.isAsync || dep.isEntry) {
-            let isIsolated = dep.isEntry || dep.env.isIsolated();
-            let resolved = assetGraph.getDependencyResolution(dep);
-            if (!resolved) {
-              // TODO: is this right?
-              return;
-            }
+    bundleGraph.traverse((node, context) => {
```

Comment:
Technically all we traverse from the bundleGraph throughout the bundler are dependencies. Maybe this could become `traverseDependencies`?